b2/src/main/java/org/jclouds/b2/features/BucketApi.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.labs.b2.features;
+
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+
+import static org.jclouds.blobstore.attr.BlobScopes.CONTAINER;
+
+import javax.inject.Named;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+
+import org.jclouds.blobstore.attr.BlobScope;
+import org.jclouds.labs.b2.domain.Bucket;
+import org.jclouds.labs.b2.domain.BucketList;
+import org.jclouds.labs.b2.domain.BucketType;
+import org.jclouds.labs.b2.filters.RequestAuthorization;
+import org.jclouds.rest.annotations.MapBinder;
+import org.jclouds.rest.annotations.PayloadParam;
+import org.jclouds.rest.annotations.PayloadParams;
+import org.jclouds.rest.annotations.RequestFilters;
+import org.jclouds.rest.binders.BindToJsonPayload;
+
+@RequestFilters(RequestAuthorization.class)
+@BlobScope(CONTAINER)
+public interface BucketApi {
+   @Named("b2_create_bucket")
+   @POST
+   @Path("/b2api/v1/b2_create_bucket")
+   @MapBinder(BindToJsonPayload.class)
+   @PayloadParams(keys = {"accountId"}, values = {"{jclouds.identity}"})
+   @Consumes(APPLICATION_JSON)
+   @Produces(APPLICATION_JSON)
+   Bucket createBucket(@PayloadParam("bucketName") String bucketName, @PayloadParam("bucketType") BucketType bucketType);
+
+   @Named("b2_delete_bucket")
+   @POST
+   @Path("/b2api/v1/b2_delete_bucket")
+   @MapBinder(BindToJsonPayload.class)
+   @PayloadParams(keys = {"accountId"}, values = {"{jclouds.identity}"})
+   @Consumes(APPLICATION_JSON)
+   @Produces(APPLICATION_JSON)
```

Comment:
Yes, it does not return a 404, but it actually throws a ResourceNotFoundException when the referenced bucket is not found (which in practice, for jclouds is just the same as if it had returned a 404).

What I mean, is that a common practice in jclouds is to avoid failing on delete operations if the resource to be deleted does not exist. [The S3 API](https://github.com/jclouds/jclouds/blob/master/apis/s3/src/main/java/org/jclouds/s3/S3Client.java#L207) and many others behave this way, and I think it would be good to follow that pattern here too.

If you annotate this method with `@Fallback(NullOnNotFoundOr404.class)`, then the method will not fail and just return `null` when the resource to be deleted does not exist, just like most of the other jclouds APIs. The fallback will not only take place if the response is a 404, [but also if the call generated a ResoueceNotFoundException](https://github.com/jclouds/jclouds/blob/master/core/src/main/java/org/jclouds/Fallbacks.java#L138-L140). Don't be "distracted" by the "Or404" in the fallback name; it checks for exceptions too, and your error handler already takes care of properly setting it when appropriate, so APIs can use this kind of callbacks.